pkg/network/master/subnets.go

```diff
@@ -19,92 +19,67 @@ import (
+	if err := master.initSubnetAllocators(); err != nil {
+		return err
 	}
-	for _, sub := range subnets {
-		if err = master.networkInfo.ValidateNodeIP(sub.HostIP); err != nil {
-			// Don't error out; just warn so the error can be corrected with 'oc'
-			utilruntime.HandleError(fmt.Errorf("Failed to validate HostSubnet %s: %v", common.HostSubnetToString(sub), err))
-		} else {
-			glog.Infof("Found existing HostSubnet %s", common.HostSubnetToString(sub))
-			_, subnetIP, err := net.ParseCIDR(sub.Subnet)
-			if err != nil {
-				return fmt.Errorf("failed to parse network address: %q", sub.Subnet)
-			}
 
-			for _, cn := range clusterNetworks {
-				if cn.ClusterCIDR.Contains(subnetIP.IP) {
-					subrange[cn] = append(subrange[cn], sub.Subnet)
-					break
-				}
-			}
-		}
-	}
-	var subnetAllocatorList []*SubnetAllocator
-	for _, cn := range clusterNetworks {
-		subnetAllocator, err := NewSubnetAllocator(cn.ClusterCIDR.String(), cn.HostSubnetLength, subrange[cn])
+	master.watchNodes()
+	master.watchSubnets()
+	return nil
+}
+
+func (master *OsdnMaster) initSubnetAllocators() error {
+	for _, cn := range master.networkInfo.ClusterNetworks {
+		sa, err := NewSubnetAllocator(cn.ClusterCIDR.String(), cn.HostSubnetLength, nil)
 		if err != nil {
 			return err
 		}
-		subnetAllocatorList = append(subnetAllocatorList, subnetAllocator)
+		master.subnetAllocatorList = append(master.subnetAllocatorList, sa)
+		master.subnetAllocatorMap[cn] = sa
 	}
-	master.subnetAllocatorList = subnetAllocatorList
+	return nil
+}
 
-	err = master.reconcileHostSubnets(subnets)
-	if err != nil {
-		return err
+func (master *OsdnMaster) watchNodes() {
+	funcs := common.InformerFuncs(&kapi.Node{}, master.handleAddOrUpdateNode, master.handleDeleteNode)
+	master.nodeInformer.Informer().AddEventHandler(funcs)
+}
+
+func (master *OsdnMaster) handleAddOrUpdateNode(obj, _ interface{}, eventType watch.EventType) {
+	node := obj.(*kapi.Node)
+
+	nodeIP := getNodeInternalIP(node)
+	if len(nodeIP) == 0 {
+		utilruntime.HandleError(fmt.Errorf("Node IP is not set for node %s, skipping %s event, node: %v", node.Name, eventType, node))
+		return
 	}
 
-	master.watchNodes()
-	master.watchSubnets()
-	return nil
-}
+	if oldNodeIP, ok := master.hostSubnetNodeIPs[node.UID]; ok && (nodeIP == oldNodeIP) {
+		return
+	}
+	// Node status is frequently updated by kubelet, so log only if the above condition is not met
+	glog.V(5).Infof("Watch %s event for Node %q", eventType, node.Name)
 
-// reconcileHostSubnets verifies and corrects the state of the hostsubnets on startup of the master.
-// Because openshift watches on events to keep hostsubnets and nodes in the correct state, missing an event
-// can cause orphaned or unusable hostsubnets to stick around.
-func (master *OsdnMaster) reconcileHostSubnets(hostSubnets []*networkapi.HostSubnet) error {
-	nodes, err := master.nodeInformer.Lister().List(labels.Everything())
+	master.clearInitialNodeNetworkUnavailableCondition(node)
```

Comment:
clearInitialNodeNetworkUnavailableCondition really ought to be somewhere else. master.go maybe. (It was only here before because we didn't want to create a second EventQueue, but with SharedInformers that's no longer an issue.)